Framework/Crystal/src/PredictSatellitePeaks.cpp

```diff
@@ -0,0 +1,293 @@
+                  "single-crystal peaks.");
+  declareProperty(make_unique<PropertyWithValue<double>>("MinDSpacing", 0.1,
+                                                         Direction::Input),
+                  "Minimum d-spacing of peaks to consider. Default = 1.0");
+  declareProperty(make_unique<PropertyWithValue<double>>("MaxDSpacing", 100.0,
+                                                         Direction::Input),
+                  "Maximum d-spacing of peaks to consider");
+
+  setPropertySettings(
+      "WavelengthMin",
+      Kernel::make_unique<Kernel::EnabledWhenProperty>(
+          string("IncludeAllPeaksInRange"), Kernel::IS_EQUAL_TO, "1"));
+
+  setPropertySettings(
+      "WavelengthMax",
+      Kernel::make_unique<Kernel::EnabledWhenProperty>(
+          string("IncludeAllPeaksInRange"), Kernel::IS_EQUAL_TO, "1"));
+  setPropertySettings(
+      "MinDSpacing",
+      Kernel::make_unique<Kernel::EnabledWhenProperty>(
+          string("IncludeAllPeaksInRange"), Kernel::IS_EQUAL_TO, "1"));
+
+  setPropertySettings(
+      "MaxDSpacing",
+      Kernel::make_unique<Kernel::EnabledWhenProperty>(
+          string("IncludeAllPeaksInRange"), Kernel::IS_EQUAL_TO, "1"));
+}
+
+/// Run the algorithm
+void PredictSatellitePeaks::exec() {
+  PeaksWorkspace_sptr Peaks = getProperty("Peaks");
+  if (!Peaks)
+    throw std::invalid_argument(
+        "Input workspace is not a PeaksWorkspace. Type=" + Peaks->id());
+
+  vector<double> offsets1 = getProperty("ModVector1");
+  vector<double> offsets2 = getProperty("ModVector2");
+  vector<double> offsets3 = getProperty("ModVector3");
+  int maxOrder = getProperty("MaxOrder");
+  if (offsets1.empty()) {
+    offsets1.push_back(0.0);
+    offsets1.push_back(0.0);
+    offsets1.push_back(0.0);
+  }
+  if (offsets2.empty()) {
+    offsets2.push_back(0.0);
+    offsets2.push_back(0.0);
+    offsets2.push_back(0.0);
+  }
+  if (offsets3.empty()) {
+    offsets3.push_back(0.0);
+    offsets3.push_back(0.0);
+    offsets3.push_back(0.0);
+  }
+
+  bool includePeaksInRange = getProperty("IncludeAllPeaksInRange");
+  bool includeOrderZero = getProperty("IncludeIntegerHKL");
+
+  if (Peaks->getNumberPeaks() <= 0) {
+    g_log.error() << "There are No peaks in the input PeaksWorkspace\n";
+    return;
+  }
+
+  API::Sample sample = Peaks->sample();
+
+  const auto &lattice = sample.getOrientedLattice();
+
+  const auto instrument = Peaks->getInstrument();
+
+  auto OutPeaks = boost::dynamic_pointer_cast<IPeaksWorkspace>(
+      WorkspaceFactory::Instance().createPeaks());
+  OutPeaks->setInstrument(instrument);
+
+  V3D hkl;
+  int peakNum = 0;
+  const auto NPeaks = Peaks->getNumberPeaks();
+  Kernel::Matrix<double> goniometer;
+  goniometer.identityMatrix();
+
+  const double lambdaMin = getProperty("WavelengthMin");
+  const double lambdaMax = getProperty("WavelengthMax");
+  IPeak &peak0 = Peaks->getPeak(0);
+
+  std::vector<V3D> possibleHKLs;
+  if (includePeaksInRange) {
```

Comment:
Looking closer at this, I think you can simplify the way this works. Fundamentally I think we have two cases:

 - Predict peaks using the `HKLGenerator`
 - Predict peaks using the peaks in the peaks workspace.

Now that we've refactored the core part of the algorithm into the `predictOffsets` function I think it might be an idea to have two separate functions for the cases above. There might be a little code duplication but it would be much clearer to read what is happening.